### PR TITLE
Feat/valkey vector store

### LINKED ISSUE #10511: [FEATURE]: Add Valkey as a vector store provider for codebase indexing
### Feature hasn't been suggested before.

- [x] I have verified this feature I'm about to request hasn't been suggested before.

### Describe the enhancement you want to request

Add Valkey (with ValkeySearch module) as a supported vector store backend for codebase indexing, alongside the existing LanceDB and Qdrant options.

**Why add another vector store?**

The current options each have friction points that Valkey addresses:

- **Qdrant** requires deploying and managing a dedicated vector database that serves no other purpose in most stacks. Many teams won't approve adding a single-purpose service just for code search.
- **LanceDB** is embedded/file-based, which limits it to single-machine use and makes it unsuitable for shared or remote development environments.

Valkey (the community-driven Linux Foundation governed Redis fork) is already deployed in a huge number of organizations for caching, session storage, and pub/sub. Its ValkeySearch module adds vector similarity search capabilities to an existing service. This means teams can enable codebase indexing without provisioning any new infrastructure — they just use what's already running.

**Use case:**

Organizations with strict infrastructure policies or privacy requirements benefit the most. They already have Valkey/Redis running, they control where their data lives, and they don't need to onboard a new service or send embeddings to a third party. Individual developers who run Valkey locally also get a zero-setup path to indexing.

Supporting other vector databases also was raised in the discussion on #6144, where a contributor asked about supporting additional vector databases beyond LanceDB and Qdrant.

**What this would involve:**

- A new `"valkey"` option for the `vectorStoreProvider` configuration field
- Connection configuration (URL + optional password) similar to how Qdrant is configured today
- Full feature parity with the existing vector store interface: initialization, upsert, cosine similarity search with directory filtering, deletion, collection lifecycle, and metadata/indexing state management
- The same graceful behavior users expect (non-existent collections don't throw, indexing state is tracked, embedding profile changes trigger re-indexing)

**Benefits:**

- Reuse existing infrastructure — no new services to deploy, monitor, or pay for
- Familiar operational tooling (valkey-cli, redis-cli, existing monitoring, backups)
- Privacy-friendly — embeddings stay on infrastructure the team already controls
- High performance — Valkey is optimized for low-latency in-memory operations
- Broad ecosystem — works with any Valkey-compatible deployment (AWS ElastiCache, self-hosted, etc.)

### PR DESCRIPTION
**Pull request by rileydes-improving**
## Issue

Fixes #10511

## Context

Adds Valkey (with ValkeySearch module) as a new vector store backend for codebase indexing, alongside LanceDB and Qdrant. Organizations already running Valkey/Redis can enable indexing without provisioning new infrastructure.

Required versions: Valkey 8.1.1+, Valkey Search 1.0+, `@valkey/valkey-glide` 2.4.0.

## Implementation

New `ValkeyVectorStore` class (752 lines) implementing `IVectorStore` via the `@valkey/valkey-glide` client and its `GlideFt` module.

Key decisions reviewers should note:

- **TAG fields over full-text search**: 5 path-segment TAG fields (`seg0`–`seg4`) enable directory-prefix filtering without the overhead of a full-text index
- **TAG sanitization**: Escapes ValkeySearch metacharacters and rejects `=>` to prevent query injection
- **Metadata hash**: A dedicated `__metadata__` key stores embedding profile and indexing-complete state, enabling profile-mismatch detection and automatic re-indexing

Config/UI integration adds `valkeyUrl`, `valkeyPassword`, and `"valkey"` to the `vectorStoreProvider` union, with a corresponding factory branch and settings UI fields.

## Screenshots / Video

| before | after |
|---|---|
| N/A — new feature | Valkey appears as a vector store option in Settings → Indexing with URL/password fields |
Screenshot 2026-05-26 at 13 58 18
Screenshot 2026-05-26 at 13 59 18

### Manual/local verification

- Selected "Valkey" in Settings → Indexing, entered `redis://localhost:6379`, confirmed indexing completed and search returned correct results with directory prefix filtering (human)

### Reviewer test steps

1. `docker run -d -p 6379:6379 valkey/valkey-bundle:latest`
2. Settings → Indexing → select "Valkey", enter `redis://localhost:6379`
3. Enable indexing, confirm it completes and search returns results
4. Run committed test suites:
   ```bash
   cd packages/kilo-indexing
   # Unit tests (mocked, no server needed)
   bun test test/kilocode/indexing/vector-store/valkey-vector-store.unit.test.ts --timeout 30000
   bun test test/kilocode/indexing/vector-store/valkey-vector-store.operations.test.ts --timeout 30000
   bun test test/kilocode/indexing/vector-store/valkey-vector-store.lifecycle.test.ts --timeout 30000
   # Property tests (no server needed)
   bun test test/kilocode/indexing/vector-store/valkey-vector-store.property.test.ts --timeout 60000
   # Integration tests (requires running Valkey with ValkeySearch)
   VALKEY_URL=redis://localhost:6379 bun test test/kilocode/indexing/vector-store/valkey-vector-store.integration.test.ts --timeout 60000
   ```

### Blocked checks and substitute verification

- Integration tests require Valkey 8.1.1+ with ValkeySearch module, which CI does not have
- Substitute: full `test-valkey-backend.sh` (10 phases) run against a Finch container — all phases passed (human tested/verified). Covers unit tests, property tests, integration tests against live Valkey, backend HTTP tests (13 sub-tests), direct CLI verification, SDK-level probe (19 assertions), password-auth test, E2E orchestrator smoke test, and typecheck
- Manual UI testing of indexing settings and workflow against the same instance (human tested/verified)

## Checklist

- [x] Issue linked above, or exception explained
- [x] Tests/verification described
- [x] Screenshots/video included for visual changes, or marked N/A
- [x] Changeset considered for user-facing changes
- [x] I personally reviewed the diff and can explain the changes, including any AI-assisted work.

### CODE REVIEW

**Comment by rileydes-improving** [comment]
Just rebased on top of the most recent main, tested, and pushed to keep it up to date.

**Comment by rileydes-improving** [comment]
I just rebased to the most recent main to ensure I am compatible with the stable indexing.

File: packages/kilo-indexing/package.json [hunk 1]
```diff
@@ -57,6 +58,7 @@
     "@tsconfig/node22": "catalog:",
     "@types/node": "catalog:",
     "@typescript/native-preview": "catalog:",
+    "fast-check": "4.7.0",
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/package.json]: Fixed, `fast-check` pinned to `4.8.0` in package.json, satisfying Effect's `^4.8.0` requirement. See `c7cd6114`.

File: packages/opencode/src/kilocode/components/dialog-indexing.tsx [hunk 2]
```diff
@@ -395,6 +398,48 @@ async function showQdrantSettings(
...
+    dialog.replace(() => <DialogIndexing useSDK={useSDK} scope={scope} />)
+    return
+  }
+
+  const updated: IndexingConfig = {
+    ...raw,
+    vectorStore: "valkey",
+    valkey: {
+      url: url.trim() || undefined,
+      password: password.trim() || undefined,
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/opencode/src/kilocode/components/dialog-indexing.tsx]: Fixed, Uses `raw.valkey?.password` (project-local only) for inheritance-safe persistence. Global passwords are never copied into project config on confirmation. See 7c175fe.

File: packages/opencode/src/kilocode/components/dialog-indexing.tsx [hunk 3]
```diff
@@ -395,6 +398,48 @@ async function showQdrantSettings(
...
+    value: currentSettings.url ?? "",
+    placeholder: "redis://localhost:6379",
+  })
+  if (url === null) {
+    dialog.replace(() => <DialogIndexing useSDK={useSDK} scope={scope} />)
+    return
+  }
+
+  const password = await DialogPrompt.show(dialog, "Valkey — Password", {
+    value: currentSettings.password ?? "",
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/opencode/src/kilocode/components/dialog-indexing.tsx]: Fixed, Password prompt uses `value: ""` with a masked placeholder (`•••`). Stored credential is never rendered on screen. See 7c175fe.

File: packages/kilo-vscode/webview-ui/src/components/settings/IndexingTab.tsx [hunk 4]
```diff
@@ -514,6 +517,39 @@ const IndexingTab: Component = () => {
             </SettingsRow>
           </>
         </Show>
+        <Show when={vectorStore() === "valkey"}>
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-vscode/webview-ui/src/components/settings/IndexingTab.tsx:617]: Fixed, LanceDB settings are guarded by `<Show when={vectorStore() === "lancedb"}>` in the fallback. Panels are now mutually exclusive. See 7c175fe.

File: packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts [hunk 5]
```diff
@@ -0,0 +1,811 @@
...
+  async dispose(): Promise<void> {
+    if (this.connectingPromise) {
+      try {
+        await this.connectingPromise
+      } catch (err) {
+        log.debug("dispose: ignored connection error", { err })
+      }
+    }
+    if (this.client) {
+      this.client.close()
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts:633]: Fixed, `close()` implemented, delegates to `dispose()`. Production lifecycle paths now correctly invoke cleanup. See 7c175fe.

File: packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts [hunk 6]
```diff
@@ -0,0 +1,811 @@
...
+      const client = await this.ensureConnected()
+      const keysToDelete: string[] = []
+      const DELETE_BATCH_SIZE = 1000
+      const SEARCH_LIMIT = 10000
+
+      for (const filePath of filePaths) {
+        const sanitizedPath = this.sanitizeTagValue(filePath)
+        const query = `@filePath:{${sanitizedPath}}`
+
+        const [, documents] = await GlideFt.search(client, this.collectionName, query, {
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts]: Fixed, Deletion is now paginated: re-queries from offset 0 in a loop until fewer than SEARCH_LIMIT results remain. See 7c175fe.

File: packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts [hunk 7]
```diff
@@ -0,0 +1,811 @@
...
+          }
+        }
+      }
+    }
+
+    return 0
+  }
+
+  private parseNumDocsFromInfo(info: Record<string, any>): number {
+    const numDocs = info.num_docs
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts:161]: Fixed, `parseNumDocsFromInfo()` handles both number and string types with `Number.isFinite()` guard. See 7c175fe.

File: packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts [hunk 8]
```diff
@@ -0,0 +1,811 @@
...
+    this.collectionName = `ws-${hash.substring(0, 16)}`
+    this.metadataKey = `${this.collectionName}:__metadata__`
+  }
+
+  async initialize(): Promise<boolean> {
+    try {
+      await this.ensureConnected()
+    } catch (error) {
+      log.error("Failed to connect to Valkey", {
+        url: this.valkeyUrl,
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts]: Fixed, `redactedUrl()` helper strips credentials before logging. All log/error paths use the redacted form. See 7c175fe.

File: packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts [hunk 9]
```diff
@@ -0,0 +1,811 @@
...
+  model: "embedding_model_id",
+  dimension: "embedding_dimension",
+}
+
+/**
+ * Valkey implementation of the vector store interface using ValkeySearch commands.
+ * Uses the @valkey/valkey-glide client library with its GlideFt module for
+ * ValkeySearch operations (FT.CREATE, FT.SEARCH, FT.DROPINDEX, FT.INFO).
+ */
+export class ValkeyVectorStore implements IVectorStore {
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/vector-store/valkey-vector-store.ts:24]: Fixed, `openExisting()` implemented with dimension, profile, and completeness validation. See 7c175fe.

File: packages/kilo-indexing/src/indexing/service-factory.ts [hunk 10]
```diff
@@ -194,6 +195,23 @@ export class CodeIndexServiceFactory {
...
+    if (config.vectorStoreProvider === "valkey") {
+      if (!config.valkeyUrl) throw new Error("Valkey URL is required.")
+      log.info("creating vector store", {
+        provider: config.embedderProvider,
+        vectorStore: "valkey",
+        model: profile.modelId,
+        vectorSize: profile.dimension,
+      })
+      return new ValkeyVectorStore(
+        this.workspacePath,
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/service-factory.ts]: Fixed, `createVectorStore(workspacePath)` passes the argument through to `ValkeyVectorStore` constructor, which hashes it for the collection name. Matches LanceDB/Qdrant behavior. See 7c175fe.

File: packages/kilo-indexing/src/indexing/service-factory.ts [hunk 11]
```diff
@@ -15,6 +15,7 @@ import { OpenRouterEmbedder } from "./embedders/openrouter"
 import { VoyageEmbedder } from "./embedders/voyage"
 import { QdrantVectorStore } from "./vector-store/qdrant-client"
 import { LanceDBVectorStore } from "./vector-store/lancedb-vector-store"
+import { ValkeyVectorStore } from "./vector-store/valkey-vector-store"
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/service-factory.ts]: Fixed, Valkey implementation is lazy-loaded via `require()` only when the Valkey backend is selected. Additionally, `@valkey/valkey-glide@2.4.0` includes Windows native binaries (support added in v2.1). See c7cd6114.

**Comment by rileydes-improving** [comment]
Rebased upon a recent main to resolve conflicts

File: packages/kilo-indexing/src/indexing/service-factory.ts [hunk 12]
```diff
@@ -293,7 +313,7 @@ export class CodeIndexServiceFactory {
 
     const embedder = this.createEmbedder()
     const vectorStore = this.createVectorStore()
-    const parser = new CodeParser(config.fileExtensions)
+    const parser = codeParser
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-indexing/src/indexing/service-factory.ts]: Addressed, restored main's version

File: packages/kilo-vscode/webview-ui/src/components/settings/IndexingTab.tsx [hunk 13]
```diff
@@ -609,6 +614,39 @@ const IndexingTab: Component = () => {
...
+          <>
+            <SettingsRow title="Valkey URL" description="Connection URL for the Valkey vector store">
```
[the remark this answers is not in the document]
Comment (rileydes-improving) [reply packages/kilo-vscode/webview-ui/src/components/settings/IndexingTab.tsx]: Added localization, modified Qdrant's localization where possible

**Comment by rileydes-improving** [comment]
@marius-kilocode Could I get a review on this branch? Let me know if there is anything that needs to be changed

**Comment by marius-kilocode** [comment]
Tested this branch against a live Valkey (`valkey/valkey-bundle`), plus the full suite, typecheck, and the VS Code settings UI. The feature works end to end. Two things block the merge, two are non-blocking.

**Blocking: Windows.** `@valkey/valkey-glide` only publishes darwin/linux native binaries, and its loader throws at import when no binding exists. The integration test imports it statically (`valkey-vector-store.integration.test.ts:12`), and the Windows CI shard runs `test:ci` for `@kilocode/kilo-indexing`, so that job goes red once this merges. Can we guard the test with a platform skip or move the import into `beforeAll`? Also, what should happen at runtime on Windows when Valkey is selected? Right now it fails with a native load error. I think either hiding the option on Windows or catching the load error in `service-factory.ts` with a clear "not supported on Windows" message is fine, but we need one of them.

**Blocking: integration test flushes the whole server.** `beforeAll`/`beforeEach` call `FLUSHDB` against the default `redis://localhost:6379` (`:18`, `:60-67`, `:93`, `:115`). On my machine that wiped the local Redis DB 0 before the tests failed with `unknown command 'FT.INFO'`. The check at `:83` treats any `RequestError` as "module loaded", so a plain Redis is misdetected as available. A plain `bun test` in the package should never destroy developer data. I would prefer an explicit opt-in env var (skip unless `VALKEY_URL` is set), an availability check that looks at the error text or `MODULE LIST`, and cleanup by key prefix instead of `FLUSHDB`. The other stores only have mocked tests, so this is the first test in the package that hits a real server.

**Non-blocking: redactedUrl edge case.** `redactedUrl()` (`valkey-vector-store.ts:54-63`) leaks part of the password when it contains an unescaped `@`: `redis://default:P@ss/w0rd@host:6379` becomes `redis://ss/w0rd@host:6379`. Building it from protocol/host/port only would avoid that. To be fair, this is already better than the Qdrant store, which logs and throws the raw URL.

**Non-blocking: inherited secrets written to project config.** In the VS Code tab the password field reads the merged value and commits it on blur (`IndexingTab.tsx:645`, `:651-654`), and in the TUI the URL is persisted from the merged config while the password is guarded (`dialog-indexing.tsx:411-445`). A global secret can end up in a project `kilo.json`. This is pre-existing behavior, Qdrant has the same problem in both UIs, so I do not want to block this PR on it. I will open a follow-up issue covering all providers. If you want to extend the TUI password guard to the URL here, fine, but not required.

One thing I checked and am not counting: unescaped `$` in `sanitizeTagValue` looked like a TAG injection risk, but it does not reproduce against a real ValkeySearch server, deletes and prefix filters behave correctly.

Also worth another rebase before merge, since main moved and `i18n/en.ts` and `openapi.json` changed a lot since the merge base.
